File: src/main/java/ru/ezhov/rocket/action/types/template/PanelEngine.java
package ru.ezhov.rocket.action.types.template;

import javax.swing.JLabel;
import javax.swing.JPanel;
import javax.swing.JTextField;
import java.awt.BorderLayout;
import java.awt.Color;
import java.awt.Cursor;
import java.awt.Desktop;
import java.awt.Graphics;
import java.awt.GridLayout;
import java.awt.event.KeyListener;
import java.awt.event.MouseAdapter;
import java.awt.event.MouseEvent;
import java.net.URI;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

class PanelEngine extends JPanel {
    private List<String> words;
    private Map<String, JTextField> map = new HashMap<>();

    public PanelEngine(List<String> words, KeyListener keyListenerForAllTextFields) {
        final ArrayList<String> strings = new ArrayList<>(words);

        Collections.sort(strings);

        this.words = strings;
        setLayout(new BorderLayout());

        JPanel panelTextField = new JPanel(new GridLayout(words.size(), 1));

        panelTextField.setLayout(new GridLayout(words.size(), 1));

        for (String s : this.words) {
            JTextField textField = new TextFieldWithText(s);
            map.put(s, textField);
            panelTextField.add(textField);
            if (keyListenerForAllTextFields != null) {
                textField.addKeyListener(keyListenerForAllTextFields);
            }
        }

        JLabel label = new JLabel("https://velocity.apache.org/engine/1.7/user-guide.html");
        final Color foregroundDefault = label.getForeground();
        final Cursor cursorDefault = label.getCursor();
        label.addMouseListener(new MouseAdapter() {
            @Override
            public void mouseReleased(MouseEvent e) {
                try {
                    Desktop.getDesktop().browse(new URI(label.getText()));
                } catch (Exception ioException) {
                    //no matter
                }
            }

            @Override
            public void mouseEntered(MouseEvent e) {
                label.setForeground(Color.BLUE);
                label.setCursor(new Cursor(Cursor.HAND_CURSOR));
            }

            @Override
            public void mouseExited(MouseEvent e) {
                label.setForeground(foregroundDefault);
                label.setCursor(cursorDefault);
            }
        });

        add(label, BorderLayout.NORTH);
        add(panelTextField, BorderLayout.CENTER);
    }

    public Map<String, String> apply() {
        return map
                .entrySet()
                .stream()
                .collect(
                        Collectors.toMap(
                                k -> k.getKey().replaceAll("\\$", ""),
                                en -> {
                                    if ("".equals(en.getValue().getText())) {
                                        return en.getKey();
                                    } else {
                                        return en.getValue().getText();
                                    }
                                }
                        )
                );
    }

    public void initVariables(Map<String, String> map) {
        map.forEach((k, v) -> {
            final JTextField textField = this.map.get("$" + k);
            if (textField != null) {
                textField.setText(v);
            }
        });
    }

    private static class TextFieldWithText extends JTextField {
        private final String text;

        public TextFieldWithText(String text) {
            this.text = text;
        }

        @Override
        public void paint(Graphics g) {
            super.paint(g);
            if ("".equals(this.getText())) {
                g.setColor(Color.gray);
                g.drawString(text, 5, this.getHeight() - 5);
            }
        }
    }
}

File: src/main/kotlin/ru/ezhov/rocket/action/types/template/PanelEngine.kt
package ru.ezhov.rocket.action.types.template

import ru.ezhov.rocket.action.ui.swing.common.TextFieldWithText
import java.awt.BorderLayout
import java.awt.Color
import java.awt.Cursor
import java.awt.Desktop
import java.awt.GridLayout
import java.awt.event.KeyListener
import java.awt.event.MouseAdapter
import java.awt.event.MouseEvent
import java.net.URI
import java.util.stream.Collectors
import javax.swing.JLabel
import javax.swing.JPanel
import javax.swing.JTextField

internal class PanelEngine(words: List<String>, keyListenerForAllTextFields: KeyListener?) : JPanel() {
    private val words: List<String>
    private val map: MutableMap<String?, JTextField> = HashMap()
    fun apply(): Map<String, String> {
        return map
            .entries
            .stream()
            .collect(
                Collectors.toMap<Map.Entry<String?, JTextField>, String, String>(
                    { (key): Map.Entry<String?, JTextField> -> key!!.replace("\\$".toRegex(), "") },
                    { (key, value): Map.Entry<String?, JTextField> ->
                        if ("" == value.text) {
                            key
                        } else {
                            value.text
                        }
                    }
                )
            )
    }

    fun initVariables(map: Map<String, String?>) {
        map.forEach { (k: String, v: String?) ->
            val textField = this.map["$$k"]
            if (textField != null) {
                textField.text = v
            }
        }
    }

    init {
        this.words = words.sortedBy { it }
        layout = BorderLayout()
        val panelTextField = JPanel(GridLayout(words.size, 1))
        panelTextField.layout = GridLayout(words.size, 1)
        for (s in this.words) {
            val textField: JTextField = TextFieldWithText(s)
            map[s] = textField
            panelTextField.add(textField)
            keyListenerForAllTextFields?.let {
                textField.addKeyListener(keyListenerForAllTextFields)
            }
        }
        val label = JLabel("https://velocity.apache.org/engine/1.7/user-guide.html")
        val foregroundDefault = label.foreground
        val cursorDefault = label.cursor
        label.addMouseListener(object : MouseAdapter() {
            override fun mouseReleased(e: MouseEvent) {
                try {
                    Desktop.getDesktop().browse(URI(label.text))
                } catch (ioException: Exception) {
                    //no matter
                }
            }

            override fun mouseEntered(e: MouseEvent) {
                label.foreground = Color.BLUE
                label.cursor = Cursor(Cursor.HAND_CURSOR)
            }

            override fun mouseExited(e: MouseEvent) {
                label.foreground = foregroundDefault
                label.cursor = cursorDefault
            }
        })
        add(label, BorderLayout.NORTH)
        add(panelTextField, BorderLayout.CENTER)
    }
}